scm-plugins/scm-svn-plugin/src/main/java/sonia/scm/repository/spi/SvnFileLockCommand.java

```diff
@@ -0,0 +1,162 @@
+
+import java.time.Instant;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Optional;
+import java.util.function.Predicate;
+
+import static java.util.Collections.singletonMap;
+import static java.util.Optional.ofNullable;
+import static java.util.stream.Collectors.toList;
+import static org.tmatesoft.svn.core.auth.BasicAuthenticationManager.newInstance;
+import static sonia.scm.ContextEntry.ContextBuilder.entity;
+
+public class SvnFileLockCommand extends AbstractSvnCommand implements FileLockCommand {
+
+  private static final String LOCK_MESSAGE_PREFIX = "locked by SCM-Manager for ";
+
+  protected SvnFileLockCommand(SvnContext context) {
+    super(context);
+  }
+
+  @Override
+  public LockCommandResult lock(LockCommandRequest request) {
+    String fileToLock = request.getFile();
+    try {
+      doLock(fileToLock);
+      return new LockCommandResult(true);
+    } catch (SVNException e) {
+      throw new InternalRepositoryException(entity("File", fileToLock).in(repository), "failed to lock file", e);
+    }
+  }
+
+  private void doLock(String fileToLock) throws SVNException {
+    SVNRepository svnRepository = open();
+    String currentUser = initializeAuthentication(svnRepository);
+    getFileLock(fileToLock, svnRepository)
+      .ifPresent(lock -> {
+        throw new FileLockedException(repository.getNamespaceAndName(), lock);
+      });
+    svnRepository.lock(singletonMap(fileToLock, null), LOCK_MESSAGE_PREFIX + currentUser, false, null);
+  }
+
+  @Override
+  public UnlockCommandResult unlock(UnlockCommandRequest request) {
+    return unlock(request, any -> true);
+  }
+
+  public UnlockCommandResult unlock(UnlockCommandRequest request, Predicate<SvnFileLock> predicate) {
+    String fileToUnlock = request.getFile();
+    try {
+      doUnlock(request, predicate);
+      return new UnlockCommandResult(true);
+    } catch (SVNException e) {
+      throw new InternalRepositoryException(entity("File", fileToUnlock).in(repository), "failed to unlock file", e);
+    }
+  }
+
+  private void doUnlock(UnlockCommandRequest request, Predicate<SvnFileLock> predicate) throws SVNException {
+    String fileToUnlock = request.getFile();
+    SVNRepository svnRepository = open();
+    initializeAuthentication(svnRepository);
+    Optional<SvnFileLock> fileLock = getFileLock(fileToUnlock, svnRepository);
+    if (fileLock.isPresent()) {
+      SvnFileLock lock = fileLock.get();
+      if (!request.isForce() && !getCurrentUser().equals(lock.getUserId()) || !predicate.test(lock)) {
+        throw new FileLockedException(repository.getNamespaceAndName(), lock);
+      }
+      svnRepository.unlock(singletonMap(fileToUnlock, lock.getId()), request.isForce(), null);
+    }
+  }
+
+  @Override
+  public Optional<FileLock> status(LockStatusCommandRequest request) {
+    String file = request.getFile();
+    try {
+      return getFileLock(file, open()).map(lock -> lock);
```

Comment:
Remove this `map`

Reply:
Believe this or not, this is necessary for the type check.
You can cast `SvnFileLock` to `FileLock`, which in this case is done implicitly by `map`. But you cannot cast `Optional<SvnFileLock>` to `Optional<FileLock>`.